service/src/main/java/com/google/auto/service/processor/AutoServiceProcessor.java

```diff
@@ -243,23 +250,24 @@ private String getBinaryNameImpl(TypeElement element, String className) {
     return getBinaryNameImpl(typeElement, typeElement.getSimpleName() + "$" + className);
   }
 
-  private DeclaredType getProviderInterface(AnnotationMirror providerAnnotation) {
-
-    // The very simplest of way of doing this, is also unfortunately unworkable.
-    // We'd like to do:
-    //    ServiceProvider provider = e.getAnnotation(ServiceProvider.class);
-    //    Class<?> providerInterface = provider.value();
-    //
-    // but unfortunately we can't load the arbitrary class at annotation
-    // processing time. So, instead, we have to use the mirror to get at the
-    // value (much more painful).
-
-    Map<? extends ExecutableElement, ? extends AnnotationValue> valueIndex =
-        providerAnnotation.getElementValues();
-    log("annotation values: " + valueIndex);
-
-    AnnotationValue value = valueIndex.values().iterator().next();
-    return (DeclaredType) value.getValue();
+  /**
+   * Returns the contents of a {@code Class[]}-typed "value" field in a given {@code annotationMirror}.
+   *
+   * <p>This method is needed because directly reading the value of such a field from an
+   * AnnotationMirror throws: <pre>
+   * javax.lang.model.type.MirroredTypeException: Attempt to access Class object for TypeMirror Foo.
+   * </pre>
+   *
+   * @param annotationMirror The AnnotationMirror to read fields from, e.g. {@link AutoService}.
```

Comment:
I don't think these tags are necessary either. The first one is obvious to me and the second, and the second one is out of date at least. The first sentence of the javadoc makes both of them clear.

Reply:
dc43d90